File: version4j/src/main/java/com/only4/algorithm/version4j/leetcode/GetIntersectionNode.java
package com.only4.algorithm.version4j.leetcode;

import com.only4.algorithm.version4j.extra.ListNode;

/**
 * @author zhenyu.jiang
 */
public class GetIntersectionNode {
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        if (headA == null || headB == null) return null;

        ListNode pointerA = headA;
        ListNode pointerB = headB;

        // 当两个指针不相等时继续遍历
        while (pointerA != pointerB) {
            // 当指针A到达末尾时，重定向到链表B的头部
            pointerA = (pointerA != null) ? pointerA.next : headB;
            // 当指针B到达末尾时，重定向到链表A的头部
            pointerB = (pointerB != null) ? pointerB.next : headA;
        }

        // 返回相交节点（如果不存在则为null）
        return pointerA;
    }
}

File: version4k/src/main/kotlin/com/only4/algorithm/version4k/leetcode/getIntersectionNode.kt
package com.only4.algorithm.version4k.leetcode

import com.only4.algorithm.version4k.extra.ListNode

fun getIntersectionNode(headA: ListNode?, headB: ListNode?): ListNode? {
    if (headA == null || headB == null) return null

    var pointerA = headA
    var pointerB = headB

    while (pointerA !== pointerB) {
        // 当pointerA到达链表末尾时，重定向到链表B的头部
        pointerA = if (pointerA == null) headB else pointerA.next

        // 当pointerB到达链表末尾时，重定向到链表A的头部
        pointerB = if (pointerB == null) headA else pointerB.next
    }

    // 返回相交节点（如果不存在则为null）
    return pointerA
}
